### DISCUSSION TITLE
Sonoff zbmini changing short network (NWK) address?

### DISCUSSION THREAD

**Discussion by deviantintegral** [Koenkk/zigbee2mqtt#15874] [Devices]
I had one of my two [ZBMINI switches](https://www.zigbee2mqtt.io/devices/ZBMINI.html#sonoff-zbmini) go offline recently. I was away, so there were no real changes being made to the network. There were no power outages or anything like that.

I did some investigation with wireshark, and discovered that the network address changed! z2m had it as `0x8BB3`, but it was reporting with `0x54ed` when I flipped the physical switch. This also had the side effect of knocking a few end devices that were using it as a router.

To test, I was able to stop zigbee2mqtt and edit the `database.db` file, changing the network address to the decimal equivalent of `21741`. I restarted, and the zbmini showed up soon after.

I also pulled out the database from a backup of about a month ago, and it had the original NWK of `0x8BB3`. So, I don't think this is a z2m or herdsman bug somehow corrupting the NWK.

Has anyone else ever seen any behaviour like this with the zbmini or other devices? I would expect the NWK to only change when pairing a device, and if the zbmini fully reset I'd expect it to lose the encryption keys too.

**Comment by Koenkk** [comment]
This is nothing strange, Zigbee devices can change their network address (but never their ieee address).

**Reply by deviantintegral** [reply]
Ah, makes sense.

So then, I wonder why z2m didn't detect the address change? According to the thread at https://community.nxp.com/t5/Wireless-Connectivity/Why-does-the-short-address-of-the-device-change/m-p/1022898 and the ZCL spec, there's supposed to be a set of network status frames that are sent to indicate the new address. I suppose those could be missed if an address change happens while z2m is down. Is there code in herdsman that's supposed to handle seeing the address change outside of those frames? I'm thinking I can try flipping changing the NWK in the database again, and see why z2m never updated its database.

**Reply by Koenkk** [reply]
The link status messages are not forwarded to zigbee-herdsman by the zigbee stack. But z2m updates the network address on a device announce: https://github.com/Koenkk/zigbee-herdsman/blob/0a0621159d366d81d3ef8ba5da548104ad4dd3fc/src/controller/controller.ts#L446

**Reply by deviantintegral** [reply]
Thanks. My guess is either the device never sent a a network update command (see [herdsman](https://github.com/Koenkk/zigbee-herdsman/blob/a8a2a1dec7f0c96ea17f6e29c3a0ddcca0777390/src/adapter/z-stack/znp/definition.ts#L1303-L1315), or the zigbee spec at "3.4.10 Network Update Command"), or it did but z2m was down at the time.

What I know for sure is that the packet capture I have from 2 days later, for around 45 minutes, didn't capture any network update commands.

Unfortunately I didn't have debug logging on, so I don't have a log message from z2m.

It looks like this case is supposed to be handled already at https://github.com/Koenkk/zigbee-herdsman/blob/0a0621159d366d81d3ef8ba5da548104ad4dd3fc/src/controller/controller.ts#L547-L554. I manually modified the database again, and after a minute z2m correctly fixed up the NWK. If this happens again, I'll look deeper to see if I can figure out why this wasn't working earlier.

**Reply by deviantintegral** [reply]
A quick update that this happened again to the same device. Again, I didn't have debug logging on (it figures that I turned it off yesterday!), and it was the same fix to update the short address. Since its been 4 months since this happened, and has only happened to one of the two zbminis I have installed, I think this must most likely be an issue with my network topology or the specific device.

**Reply by deviantintegral** [reply]
I continued to have this issue every 4-6 months. Yet a second one installed has worked fine for years now.

I replaced the relay with a new ZBMINI I had (it was a 4 pack) in June and so far it hasn't dropped. But it will take several more months before I can say that the problem was with that one device.

**Comment by saveriol** [comment]
Hello!
In #20045 devices change address but this is not reflected in Z2M.

> It looks like this case is supposed to be handled already at https://github.com/Koenkk/zigbee-herdsman/blob/0a0621159d366d81d3ef8ba5da548104ad4dd3fc/src/controller/controller.ts#L547-L554

So I don't think this is working.

**Reply by deviantintegral** [reply]
I've had many other devices change NWKs without issues. If you haven't gone down the rabbit hole yet, you may want to set up a zigbee sniffer so you can watch what your devices send after you restore power.

**Reply by saveriol** [reply]
@Koenkk noob question: if the device changes address but somehow Z2M does not catch the device announce (and so the network address is updated here: https://github.com/Koenkk/zigbee-herdsman/blob/0a0621159d366d81d3ef8ba5da548104ad4dd3fc/src/controller/controller.ts#L446-L450 ) how can the device be restored online? Only manually? Is there any possibility to query the device using ieeeAddr or trigger a device announce via broadcast?

**Reply by deviantintegral** [reply]
I did a deep dive into this a few months ago, and the problem is that there's no message sent by the zstack adapter that contains both the ieee long address and the new short address after the missed join message. I think the firmware would have to be changed to improve that.

**Reply by saveriol** [reply]
thanks @deviantintegral, frankly that's absurd, so the device simply stops working and nothing can be made except manually change the address in the db?

**Reply by deviantintegral** [reply]
Well, the normal way would be to just re-pair the device. Even with the zbmini you can do it with a Philips Hue remote, so it's not too bad. Realistically, you can only edit it in the database if you've sniffed the new address in Wireshark.

Every other device I've used sends messages periodically that can be used to fix things up manually. If you can find a packet that is surfaced from the adapter with both the new NWK and the IEEE address, you could possibly improve this similar to what I did over at https://github.com/Koenkk/zigbee-herdsman/pull/992. But it's entirely possible the device is just poorly made and yes, you would be down to re-pairing in that case.

**Reply by saveriol** [reply]
In your PR there is a comment which points to https://github.com/Koenkk/zigbee-herdsman/blob/aac0732de2d3f0f88df5526adc54b54c607d4917/src/adapter/z-stack/adapter/zStackAdapter.ts#L476 and that line seems to indicate it's possible to ask for network address using ieeeAddr. 

I understand that correctly? Is it something specific to z-stack adapters?

**Reply by saveriol** [reply]
Looking at z-stack documentation:

https://software-dl.ti.com/simplelink/esd/simplelink_cc13x2_26x2_sdk/4.20.00.35/exports/docs/zigbee/doxygen/zigbee/html/zstackapi_8c.html#a28ad250922afd6565876f8769bb019de

> Calling this function will generate a message to ask for the 16 bit address of the Remote Device based on its known IEEE address. This message is sent as a broadcast message to all devices in the network.

There's also in ember:

https://docs.silabs.com/d/zigbee-stack-api/7.2.2/util-zdo#ember-network-address-request

But if I understand correctly there's no implementation of this in Z2M to check if a device is offline because of a mismatching network address.

**Reply by deviantintegral** [reply]
Ooo, thank you! I certainly missed that in both cases. I can hopefully take a look - my second zbmini just changed short addresses on me, so I know this is not a single-device issue anymore. I'd rather make z2m handle this if possible than replace the relay!

**Reply by saveriol** [reply]
Yes the best would be improve https://github.com/Koenkk/zigbee-herdsman/pull/992 to ask for the new address and then update it, I also tried to have a look at it but frankly I'm lost (I'm not really a programmer, more like a sysadmin 😆)

**Reply by deviantintegral** [reply]
I filed a draft PR at https://github.com/Koenkk/zigbee-herdsman/pull/1183 that brought my zbmini back online without having to manually edit the database.

**Reply by saveriol** [reply]
Nice!!! I would test it, but I am using ember though, is it possible to port the same in here?

**Reply by deviantintegral** [reply]
I don't see even the underlying `emberNetworkAddressRequest` implementation in ember.ts, and I don't own an ember stick to test with. Someone else will have to pick that up if it's possible!

**Reply by saveriol** [reply]
I see! The ember adapter code is also very obscure for me compared to the z-stack one... 
In any case if this gets merged I'll get a z-stack adapter ;)

**Reply by Nerivec** [reply]
I haven't had that issue with over a dozen ZBMINI. Must be different firmware versions (_they could have changed the code and not the reported version..._).
Seems like the device changes address but isn't advertising it on reconnect. Does it work again if you power-cycle the ZBMini (without any manual change)?

@deviantintegral Sidenote: a lot of methods in silabs code have multiple names but point to the same place, and several are just wrappers for basic ZDO requests (like that one).

**Reply by deviantintegral** [reply]
No, it doesn't work if I power cycle the zbmini :(

Mine have firmware date `20200927`. I do have a vague memory of them having two different firmwares for it - but no published firmware updates or update support.

One difference might be that my garage light (which is where I see this problem) doesn't have a direct connection at all to the coordinator, whereas my kitchen table light (which has been fine for years) does.

**Reply by Nerivec** [reply]
Indeed mine are all `20210320`/`2.0.0`. _Though as I mentioned, I don't particularly trust that dates/versions get properly updated whenever they actually compile a new version, since these are manually set, especially for devices that don't support OTA or never (will) get firmware updates._

Can you lookup which router it is connected to (if it's one of the usual troublemakers, Tuya, Aqara...)? Might be something there, if the message gets lost, or is not passed to the coordinator... Would need sniffing to get more details though.

**Reply by deviantintegral** [reply]
When I last looked, it was routing through Hue bulbs. I've also seen it route through Enbrighten light switches. However, I think this was fixed in https://github.com/Koenkk/zigbee-herdsman/pull/1194. So far, I've had no drops!
